src/libAtomVM/term.h

```diff
@@ -1389,6 +1439,105 @@ static inline uint32_t term_get_external_pid_serial(term t)
 
     #elif TERM_BYTES == 4
         return (uint32_t) boxed_value[4];
+    #else
+        #error "terms must be either 32 or 64 bit wide"
+    #endif
+}
+
+/*
+ * @brief Make an external reference term from node, creation, number of words and words
+ *
+ * @param node name of the node (atom)
+ * @param len number of words (1..5)
+ * @param data words
+ * @param creation creation of that node
+ * @param heap the heap to allocate memory in
+ * @return an external heap term created using given parameters.
+ */
+static inline term term_make_external_reference(term node, uint16_t len, uint32_t *data, uint32_t creation, Heap *heap)
+{
+    TERM_DEBUG_ASSERT(term_is_atom(node));
+
+    term *boxed_value = memory_heap_alloc(heap, EXTERNAL_REF_SIZE(len));
+    int atom_index = term_to_atom_index(node);
+    boxed_value[0] = ((EXTERNAL_REF_SIZE(len) - 1) << 6) | TERM_BOXED_EXTERNAL_REF;
+
+    #if TERM_BYTES == 8
+        boxed_value[1] = (term) (((uint64_t) creation) << 32 | atom_index);
+        boxed_value[2] = (term) (((uint64_t) len) << 32 | data[0]);
+        for (int i = 1; i < len; i += 2) {
+            uint64_t word = ((uint64_t) data[i]) << 32;
+            if (i + 1 < len) {
+                word |= data[i + 1];
+            }
+            boxed_value[3 + ((i - 1) / 2)] = word;
+        }
+
+    #elif TERM_BYTES == 4
+        boxed_value[1] = (term) atom_index;
+        boxed_value[2] = (term) creation;
+        for (int i = 0; i < len; i++) {
+            boxed_value[3 + i] = data[i];
+        }
+
+    #else
+        #error "terms must be either 32 or 64 bit wide"
+    #endif
+
+    return ((term) boxed_value) | TERM_BOXED_VALUE_TAG;
+}
+
+/**
+ * @brief Get the number of words of an external reference
+ *
+ * @param term external term
+ * @return the number of external nodes
+ */
+static inline uint32_t term_get_external_reference_len(term t)
+{
+    TERM_DEBUG_ASSERT(term_is_external_reference(t));
+
+    const term *boxed_value = term_to_const_term_ptr(t);
+
+    #if TERM_BYTES == 8
+        return (uint32_t) (boxed_value[2] >> 32);
+
+    #elif TERM_BYTES == 4
+        return (uint32_t) (boxed_value[0] >> 6) - 2;
+
+    #else
+        #error "terms must be either 32 or 64 bit wide"
+    #endif
+}
+
+/**
+ * @brief Get the words of an external reference
+ *
+ * @param term external term
+ * @param data the words to read, must be at least len long (maximum: 5)
+ */
+static inline void term_get_external_reference_words(term t, uint32_t *data)
```

Comment:
As I mentioned elsewhere, I'm not loving this function since it forces to make a copy of the words, instead of just accessing them.
Anyway, if we want to keep it, I'm wondering if it might be better to provide `len` as a parameter so the buffer size where words are copied must be explicitly managed and known.